common/src/main/java/com/linecorp/centraldogma/common/PathPatternBuilder.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.common;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.centraldogma.common.PathPatternOptions.PathPatternOptionFactory;
+
+/**
+ * Builds a new {@link PathPattern}.
+ *
+ * <h2>Example</h2>
+ * <pre>{@code
+ * final PathPattern factory =
+ *         PathPattern.builder()
+ *                    .startsWith("/foo/bar")
+ *                    .contains("/ext")
+ *                    .extension("json")
+ *                    .build();
+ * }</pre>
+ */
+public final class PathPatternBuilder {
+
+    private final Map<String, PathPatternOption> options = new LinkedHashMap<>();
+
+    /**
+     * Adds {@link PathPatternOptions#ENDS_WITH_OPTION}.
+     */
+    public PathPatternBuilder endsWith(String filename) {
+        put(PathPatternOptions.ENDS_WITH_OPTION, filename);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#EXTENSION_OPTION}.
+     */
+    public PathPatternBuilder extension(String extension) {
+        put(PathPatternOptions.EXTENSION_OPTION, extension);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#STARTS_WITH_OPTION}.
+     */
+    public PathPatternBuilder startsWith(String dirPath) {
+        put(PathPatternOptions.STARTS_WITH_OPTION, dirPath);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#CONTAINS_OPTION}.
+     */
+    public PathPatternBuilder contains(String dirPath) {
+        put(PathPatternOptions.CONTAINS_OPTION, dirPath);
```

Comment:
We may allow multiple `contains` unlike other methods such as `startWith()` and `enddWith()`.

- `startsWith()` supports only one path. We can simply set the value as a string field.
`endsWith()` and `extension()` are mutually exclusive. We can make the methods override the same field.
```java
public final class PathPatternBulider {

    @Nullable
    private String startPattern;
    @Nullable
    private String endPattern;
    @Nullable
    private List<String> innerPatterns;

    public PathPatternBuilder startsWith(String dirPath) {
        // validation and normalize ...
        this.startPattern = dirPath;
    }


    public PathPatternBuilder extension(String extension) {
        // validation and normalize ...
        this.endPattern = normalize(extension);
    }
}
```